fedimint-load-test-tool/src/main.rs

```diff
@@ -0,0 +1,627 @@
+        result.iter().filter(|r| r.is_err()).count()
+    );
+    for r in result {
+        if let Err(e) = r {
+            warn!("Task failed: {:?}", e);
+        }
+    }
+    Ok(())
+}
+
+async fn run_economy(
+    users: usize,
+    connect: String,
+    initial_notes: TieredMulti<SpendableNote>,
+    generate_invoice_with: Option<LnType>,
+    gateway_public_key: Option<String>,
+    event_sender: mpsc::UnboundedSender<Event>,
+) -> anyhow::Result<Vec<BoxFuture<'static, anyhow::Result<()>>>> {
+    let connect_obj: WsClientConnectInfo =
+        WsClientConnectInfo::from_str(&connect).context("invalid connect info")?;
+    let api = Arc::new(WsFederationApi::from_connect_info(&[connect_obj.clone()]))
+        as Arc<dyn IFederationApi + Send + Sync + 'static>;
+    let cfg: ClientConfig = api.download_client_config(&connect_obj).await?;
+
+    let mut tg = TaskGroup::new();
+
+    let coordinator = build_client(&cfg, &mut tg).await?;
+    let mut users_clients = Vec::with_capacity(users);
+    for _ in 0..users {
+        let client = build_client(&cfg, &mut tg).await?;
+        if let Some(gateway_public_key) = &gateway_public_key {
+            switch_default_gateway(&client, gateway_public_key).await?;
+        }
+        users_clients.push(client);
+    }
+
+    // distribute notes
+    let mut amount = initial_notes.total_amount();
+    let per_user = Amount::from_msats(1024);
+
+    info!("Reissuing initial notes: initial {amount}, per user: {per_user}");
+    reissue_notes(&coordinator, initial_notes, &event_sender).await?;
+
+    info!("Reminting notes of denomination {per_user} for {users} users (this may take a while if the number of users is high)");
+    remint_denomination(&coordinator, per_user, users.try_into()?).await?;
+
+    info!("Note summary:");
+    let summary = get_note_summary(&coordinator).await?;
+    for (k, v) in summary.iter() {
+        info!("{k}: {v}");
+    }
+
+    let mut notes_per_user = Vec::new();
+    for u in 0..users {
+        let (_, notes) = do_spend_notes(&coordinator, per_user, &event_sender).await?;
+        let user_amount = notes.total_amount();
+        info!("Giving {user_amount} to user {u}");
+        notes_per_user.push(notes);
+        amount -= user_amount;
+    }
+
+    info!("Starting user tasks");
+    let futures = users_clients
+        .into_iter()
+        .zip(notes_per_user)
+        .map(|(client, notes)| {
+            let event_sender = event_sender.clone();
+            let f: BoxFuture<_> = Box::pin(do_user_task(
+                client,
+                notes,
+                generate_invoice_with,
+                event_sender,
+            ));
+            f
+        })
+        .collect::<Vec<_>>();
+
+    Ok(futures)
+}
+
+async fn do_user_task(
+    client: Client,
+    notes: TieredMulti<SpendableNote>,
+    generate_invoice_with: Option<LnType>,
+    event_sender: mpsc::UnboundedSender<Event>,
+) -> anyhow::Result<()> {
+    let amount = notes.total_amount();
```

Comment:
Shouldn't this be in some kind of a loop so each user does this things multiple times over certain time?

Reply:
Ideally yes, in practice 1) this is tricky (you can't reissue the same notes twice and it's very slow to split the initial big notes in small notes) and 2) doing once with dozens of users is enough to measure how slow is the system

Reply:
Lots of systems can sustain a burst of load initially, but then slow down significantly after caches and write buffers get full, deallocators and garbage collectors start hitting fragmentation and other resource reclamation and so and so.

Splitting notes could be the part of natural user flow maybe. E.g. the user could always send 20% of what they have, etc.

Reply:
Added a `--notes-per-user`. So the user now can reissue many notes instead of just one